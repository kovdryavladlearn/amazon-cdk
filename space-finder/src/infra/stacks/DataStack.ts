import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib'
import { AttributeType, ITable, Table } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
import { getSuffixFromStack } from '../Utils';
import { Bucket, BucketAccessControl, HttpMethods, IBucket, ObjectOwnership } from 'aws-cdk-lib/aws-s3';

export class DataStack extends Stack{
    
    public readonly spacesTable: ITable;
    public readonly photosBucket: IBucket;
    
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const suffix = getSuffixFromStack(this);

        this.photosBucket = new Bucket(this, "SpaceFinderPhotos", {
            bucketName: `space-finder-photos-${suffix}`,
            cors: [{
                allowedMethods: [
                    HttpMethods.HEAD,
                    HttpMethods.GET,
                    HttpMethods.PUT
                ],
                allowedOrigins: ['*'],
                allowedHeaders: ['*']
            }],
            //accessControl: BucketAccessControl.PUBLIC_READ //currentry doesn't work
            blockPublicAccess:{
                blockPublicAcls: false,
                blockPublicPolicy: true,
                ignorePublicAcls: false,
                restrictPublicBuckets: false,
                
            },
            objectOwnership: ObjectOwnership.OBJECT_WRITER
        })

        new CfnOutput(this, "SpaceFinderPhotosBucketName", {
            value: this.photosBucket.bucketName
        });

        this.spacesTable = new Table(this, "SpacesTable", {
            partitionKey: {
                name: "id",
                type: AttributeType.STRING
            },
            tableName: `SpaceTable-${suffix}`
        });
    }
}