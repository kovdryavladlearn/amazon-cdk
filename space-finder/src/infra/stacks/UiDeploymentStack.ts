import { CfnOutput, Stack, StackProps } from "aws-cdk-lib";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { getSuffixFromStack } from "../Utils";
import { join } from "path";
import { existsSync } from "fs";
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment";
import { AccessLevel, Distribution, OriginAccessIdentity } from "aws-cdk-lib/aws-cloudfront";
import { S3BucketOrigin } from "aws-cdk-lib/aws-cloudfront-origins";

export class UiDeploymentStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const suffix = getSuffixFromStack(this);

        console.log(suffix);


        const deploymentBucket = new Bucket(this, "uiDeploymentBucket", {
            bucketName: `space-finder-frontend-${suffix}`
        });

        const uiDir = join(__dirname, '..', '..', '..', '..', 'space-finder-frontend', 'dist');

        if(!existsSync(uiDir)){
            console.warn("Ui dir not found" + uiDir);
            return;
        }

        new BucketDeployment(this, "SpacesFinderDeployment", {
            destinationBucket: deploymentBucket,
            sources: [Source.asset(uiDir)]
        });

        const originIdentity = new OriginAccessIdentity(this, "OriginAccessIdentity");
        deploymentBucket.grantRead(originIdentity);

        const s3Origin = S3BucketOrigin.withOriginAccessControl(deploymentBucket, {
            originAccessLevels: [AccessLevel.READ],
        });

        const distribution = new Distribution(this, 'SpacesFinderDistribution', {
            defaultRootObject: 'index.html',
            defaultBehavior: {
                origin: s3Origin
            }
        });

        new CfnOutput(this, "SpaceFinderUrl", {
            value: distribution.distributionDomainName
        });
    }
}