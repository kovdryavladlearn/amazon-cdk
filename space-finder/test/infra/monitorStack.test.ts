import { App } from "aws-cdk-lib";
import { MonitorStack } from "../../src/infra/stacks/MonitorStack";
import { Capture, Match, Template } from "aws-cdk-lib/assertions";
import { Handler } from "aws-cdk-lib/aws-lambda";


describe('Initial test suite', () => {
    
    let monitorStackTemplate;

    beforeAll(()=>{
        const testApp = new App({
            outdir: 'cdk.out'
        });

        const monitorStack = new MonitorStack(testApp, "MonitorStack");
        monitorStackTemplate = Template.fromStack(monitorStack);

    })

    test('Lambda properties', () => {
        monitorStackTemplate.hasResourceProperties("AWS::Lambda::Function", {
            Handler: 'index.handler',
            Runtime: 'nodejs18.x'
        })
    });

    test('Sns topic properties', () => {
        monitorStackTemplate.hasResourceProperties("AWS::SNS::Topic", {
            DisplayName: "AlarmTopic",
            TopicName: "AlarmTopic"
        })
    });

    test('Sns subscription - with matchers' , () => {
        monitorStackTemplate.hasResourceProperties("AWS::SNS::Subscription", 
        Match.objectEquals(
            {
                Protocol: 'lambda',
                TopicArn: {
                    Ref: Match.stringLikeRegexp("AlarmTopic") 
                },
                Endpoint: {
                    'Fn::GetAtt': [
                        Match.stringLikeRegexp("webHookLambda"),
                        'Arn'
                    ]
                }
            }
        ))
    });

    test('Sns subscription - with exact values' , () => {
        const snsTopic = monitorStackTemplate.findResources("AWS::SNS::Topic");
        const snsTopicName = Object.keys(snsTopic)[0];
        
        const lambda = monitorStackTemplate.findResources("AWS::Lambda::Function");
        const lambdaName = Object.keys(lambda)[0];
        
        monitorStackTemplate.hasResourceProperties("AWS::SNS::Subscription", {
            Protocol: 'lambda',
            TopicArn: {
                Ref: snsTopicName
            },
            Endpoint: {
                'Fn::GetAtt': [
                    lambdaName,
                    'Arn'
                ]
            }
        });
    });

    test('Alarm actions' , () => {
        const alarmActionsCapture = new Capture()

        monitorStackTemplate.hasResourceProperties("AWS::CloudWatch::Alarm", {
            AlarmActions: alarmActionsCapture
        });

        expect(alarmActionsCapture.asArray()).toEqual([{
            Ref: expect.stringMatching(/^AlarmTopic/)
        }])
    });

    test('Monitor stack snapshots' , () => {
       expect(monitorStackTemplate.toJSON()).toMatchSnapshot();
    });


    test('Lambda stack snapshot', () => {
        const lambda = monitorStackTemplate.findResources("AWS::Lambda::Function");

        expect(lambda).toMatchSnapshot();
    });

    test('SnsTopic stack snapshot', () => {
        const snsTopic = monitorStackTemplate.findResources("AWS::SNS::Topic");

        expect(snsTopic).toMatchSnapshot();
    });
})