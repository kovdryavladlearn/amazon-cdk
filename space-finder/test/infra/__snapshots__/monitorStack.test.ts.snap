// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial test suite Lambda stack snapshot 1`] = `
{
  "webHookLambda3427ABA6": {
    "DependsOn": [
      "webHookLambdaServiceRole1A032560",
    ],
    "Properties": {
      "Code": {
        "S3Bucket": {
          "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
        },
        "S3Key": "27b9d9a047e91bfc3e09cb45a9bb997023cf3e136685506776aac591d5be04d8.zip",
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "webHookLambdaServiceRole1A032560",
          "Arn",
        ],
      },
      "Runtime": "nodejs18.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`Initial test suite Monitor stack snapshots 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AlarmTopicD01E77F9": {
      "Properties": {
        "DisplayName": "AlarmTopic",
        "TopicName": "AlarmTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "spacesApi4xxAlarmA15E009C": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "AlarmTopicD01E77F9",
          },
        ],
        "AlarmName": "SpacesApi4xxAlarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "SpacesApi",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "4XXError",
        "Namespace": "AWS/ApiGateway",
        "OKActions": [
          {
            "Ref": "AlarmTopicD01E77F9",
          },
        ],
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 5,
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "webHookLambda3427ABA6": {
      "DependsOn": [
        "webHookLambdaServiceRole1A032560",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "27b9d9a047e91bfc3e09cb45a9bb997023cf3e136685506776aac591d5be04d8.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "webHookLambdaServiceRole1A032560",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "webHookLambdaAlarmTopic3DF832BF": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "webHookLambda3427ABA6",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "AlarmTopicD01E77F9",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "webHookLambdaAllowInvokeMonitorStackAlarmTopic9F999C82CCFBCA23": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "webHookLambda3427ABA6",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "AlarmTopicD01E77F9",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "webHookLambdaServiceRole1A032560": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Initial test suite SnsTopic stack snapshot 1`] = `
{
  "AlarmTopicD01E77F9": {
    "Properties": {
      "DisplayName": "AlarmTopic",
      "TopicName": "AlarmTopic",
    },
    "Type": "AWS::SNS::Topic",
  },
}
`;
